README.md
# AE Modulator

Jde o program který provádí FM, MFM a RLL modulace dat tak jak je vyžadováno předmětem __Architektura Počítačů__. Moduluje
jakýkoliv ASCII string a má na výběr z obou verzí RLL modulace. Při sestavování ze zdroje si jednoduše můžete definovat svou
vlastní tabulku RLL.

Online verze se připravuje ale asi vám ji hostovat nebudu

- Postup kompletnosti
  - [x] FM modulace
  - [x] MFM modulace
  - [x] RLL1 modulace
  - [x] RLL2 modulace
  - [x] Pokrytí testy
  - [ ] Základní možnosti výstupu (možnost vyžádat si jen jednu konverzi)
  - [ ] Zachycení chyb (filtrovat ěščřžýáíéóúů a všechno mimo ASCII znaky)
  - [ ] Reverzní převod?
  - [ ] Přepsat zdroják do angličtiny?
  - [ ] Spočítat efektivitu každého převodu (počet pulzů na string)

## Instalace/použití

1. Ze zdroje
   - Nainstalujte si nějakou javu
   - Nainstalujte si clojure
   - Nainstalujte is leiningen
   - Naklonujte si repo ``git clone git@github.com:qespr/ae-utb-modulator.git``
   - Přejděte do složky ``cd ae-utb-modulator`` a spustit pomocí leiningen: ``lein run [váš text]``
2. Předsestavené
   - Stáhněte si binárku ze stránky [vydání](https://github.com/qespr/ae-utb-modulator/releases)
   - Otevřete terminál tam kde jste to stáhli
   - Pomocí ``$ java -jar utb-modulator-1.0.jar [vas text]"`` získáte na standartním výstupu všechny převody
3. Online verze (dobře to není úplně instalace)
   - Asi by to teoreticky šlo narvat do nějakého online interpreteru
   - Později to můžu zkusit podporovat
4. Emacs
   - Možná bude Elisp verze, pokud se na to nevyseru (pravděpodobně se na to vyseru)

## Ukázky

```
$ java -jar utb-modulator-1.0.jar Luna
Text:  Luna
   Bin:  01001100011101010110111001100001
    FM:  PNPPPNPNPPPPPNPNPNPPPPPPPNPPPNPPPNPPPPPNPPPPPPPNPNPPPPPNPNPNPNPP
   MFM:  PNNPNNPNNPNPNNPNPNNPNPNPNNNPNNNPNNNPNPNNNPNPNPNNPNNPNPNNPNPNPNNP
 RLL-1:  NPNNPNNNNNNNPNNNNPNNNNPNNPNNNPNNNPNNPNNPNNNNNNPNNNNNNNPNNNPNNNNNPNNN
 RLL-2:  PNNPNNNNPNNNNNNPNNPNNNNPNNNPNNNPNNPNNNNNPNNNNPNNNNPNNNNNNPNNPNNPNN
```

### Bugs

- Bug: Pokud vložíte něco mimo ASCII tak to vyplivne úplný bordel
  - Řešení: Nedělejte to

## License

Copyright © 2022 qespr

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) version 3, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-05-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-05-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/modulator/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/modulator/compare/0.1.0...0.1.1


test/modulator/core_test.clj
(ns modulator.core-test
  (:require [clojure.test :refer :all]
            [modulator.core :refer :all]
            [modulator.number-utils :refer :all]))

(deftest number-utils-test
  (testing "Binární konverze"
    (is (= "101010" (do-binarky 42)))
    (is (= "1000101" (do-binarky 69)))
    (is (= "1100001001" (do-binarky 777))))
  (testing "Zarovnávání na bajt"
    (is (= "11111111" (zarovnej-na-byte "11111111")) "Selhání už přesné délky")
    (is (= "00000101" (zarovnej-na-byte "101")) "Selhání výchozí délky")
    (is (= "0000000000101010" (zarovnej-na-byte "101010" 2)) "Selhání 16 bit délky")
    (is (= "00000000000000000000000101011100" (zarovnej-na-byte "101011100" 4)) "Selhání 32bit délky")
    (is (= "0000000000000000000000000000000000000000000000000000000110011001" (zarovnej-na-byte "110011001", 8)) "Selhání 64 délky"))
  (testing "Znak na bin"
    (is (= "01100001" (znak-to-bin \a)) "Malý znak")
    (is (= "01011010" (znak-to-bin \Z)) "velký znak")) ;;Todo přidat Test na házení chyb při znaku mimo a-Z
  (testing "Text na bin"
    (is (= "01000110011001010110110001101001011000110110100101100001" (text-to-bin "Felicia")) "Text malé i velké")
    (is (= "01001001010101110101010001000011010010010101001001000100" (text-to-bin "IWTCIRD")) "All caps text")
    (is (= "01100101011011010110000101100011011100110011010001100101011101100110010101110010" (text-to-bin "emacs4ever")) "Malá písmena + čísla")))

(deftest encode-tests
  (testing "FM modulace"
    (is (= "PPPNPPPPPNPN" (fm-encode "101100")) "Ukázková FM"))
  (testing "MFM modulace"
    (is (= "NPNNNPNPNNPN" (mfm-encode "101100")) "Ukázková MFM"))
  (testing "RLL modulace"
    (is (= "NPNNNPNNPNNNPNNNNPNNNNNNPNNNNNPNNNNNNNPNNNPNNNPNNPNNPNNPNNPNNNPNNPNNNNNNPNNNNNPNNPNNNNPNNNPNNPNN"
           (rll-encode RLL-1 "010100000111001001100001011010010111001101100101")) "RLL-1")
    (is (= "PNNPNNNNNPNNPNNNNNPNNNNNPNNPNNNPNNPNNNNNPNNNNNNPNNPNNNNNPNNPNNNPNNPNNNNPNNNNPNNNNNPNNNNPNNNPNNNNNNPNNNPNNPNNPNNPNNPNNNNNNPNNPNNPNN"
           (rll-encode RLL-2 "0100001101100101011011000110010101110011011101000110100101100001")) "RLL-2")))


src/modulator/number-utils.clj
(ns modulator.number-utils
  (:gen-class))


(defn do-binarky
  "Převede Desítkové číslo do binárky"
  ([cislo] (do-binarky cislo nil  ""))
  ([cislo zbytek ret]
   (if (> cislo 0)
     (recur (quot cislo 2) (mod cislo 2) (str zbytek ret))
     ;;Způsob jakým předáváem způsobuje že se bin čísla zapisují až v další
     ;;Iteraci, proto je potřeba provézt poslední zápis zde
     (str zbytek ret))))
 ;; todo: přepsat na test
(do-binarky 16)

(defn zarovnej-na-byte
  "Zarovná binární číslo na specifikovanou pevnou délku v bajtech"
  ([bin-str] (zarovnej-na-byte bin-str 1))
  ([bin-str na-bajt]
   (str (apply str
          ;;Vygeneruje tolik nul kolik je potřeba aby jsme
          ;;se zarovnaly na cílový bajt
          (repeat (-
                   (* 8 na-bajt)
                   (count bin-str))
                  "0"))
        bin-str)))
(zarovnej-na-byte "000") ;;todo: přepsat na test

(defn znak-to-bin
  "Převede znak na jeho binární ASCII reprezentaci ve formě stringu"
  [znak]
  (zarovnej-na-byte (do-binarky (int znak))))

(defn text-to-bin
  "Převede string na jeho binární ASCII reprezentaci"
  [text]
  (apply str (map znak-to-bin text)))


src/modulator/number_utils.clj
(ns modulator.number-utils
  (:gen-class))


(defn do-binarky
  "Převede Desítkové číslo do binárky"
  ([cislo] (do-binarky cislo nil  ""))
  ([cislo zbytek ret]
   (if (> cislo 0)
     (recur (quot cislo 2) (mod cislo 2) (str zbytek ret))
     ;;Způsob jakým předáváme způsobuje že se bin čísla zapisují až v další
     ;;Iteraci, proto je potřeba provézt poslední zápis zde
     (str zbytek ret))))

(defn zarovnej-na-byte
  "Zarovná binární číslo na specifikovanou pevnou délku v bajtech"
  ([bin-str] (zarovnej-na-byte bin-str 1))
  ([bin-str na-bajt]
   (str (apply str
          ;;Vygeneruje tolik nul kolik je potřeba aby jsme
          ;;se zarovnaly na cílový bajt
          (repeat (-
                   (* 8 na-bajt)
                   (count bin-str))
                  "0"))
        bin-str)))
(zarovnej-na-byte "000") ;;todo: přepsat na test

(defn znak-to-bin
  "Převede znak na jeho binární ASCII reprezentaci ve formě stringu"
  [znak]
  (zarovnej-na-byte (do-binarky (int znak))))

(defn text-to-bin
  "Převede string na jeho binární ASCII reprezentaci"
  [text]
  (apply str (map znak-to-bin text)))


src/modulator/core.clj
(ns modulator.core
  (:require [modulator.number-utils :as nu]
            [clojure.string :as cstr])
  (:gen-class))

(def RLL-1
  "První verze RLL modulace jak je definováno v návodu na cvičení"
  {"00" "PNNN",
   "01" "NPNN",
   "100" "NNPNNN",
   "101" "PNNPNN",
   "111" "NNNPNN",
   "1100" "NNNNPNNN",
   "1101" "NNPNNPNN"})

(def RLL-2
  "Druhá verze RLL modulace jak je definováno v návodu na cvičení"
  {"11" "PNNN",
   "10" "NPNN",
   "011" "NNPNNN",
   "010" "PNNPNN",
   "000" "NNNPNN",
   "0010" "NNPNNPNN",
   "0011" "NNNNPNNN"})

(defn- rll-match
  "Zjistí který segment se nachází jako další v binárním řetězci
  Vrací segment, takže si buď můžete udělat substring pomocí jeho délky;
  nebo jej přímo použít pro lookup"
  ([klice bin-string] (rll-match klice klice bin-string))
  ;;obsahuje navíc kopii klíčů [kk] aby jsme mohli obnovit hledání po přidání nul
  ([klice kk bin-string]
  ;;Je potřeba pohlídat jestli klíče nejsou nenenulové, pak se přidávají až dvě nuly
  (cond
    ;;pokud jsou klíče už prázdné - žádná kombinace se nenašla, musíme přidávat nuly;
    ;;můžou se přidat max 2 nuly ale zde na to sereme, zatím to funguje
    (= klice '()) (recur kk kk (str bin-string "0"))
    (cstr/starts-with? bin-string (first klice)) (first klice)
    :else (recur (rest klice) kk bin-string))))


(defn rll-encode
  "Zakóduje binární string pomocí specifikované RLL tabulky"
  ([tabulka bin-string] (rll-encode tabulka bin-string nil))
  ([tabulka string-rest ret]
   (if (> (count string-rest) 0)
     (let [mezi-klic (rll-match (keys tabulka) string-rest)]
     (recur
      tabulka
      (if (< (count mezi-klic) (count string-rest))
        (subs string-rest (count mezi-klic)) ;;Pokračuje v rekzurzi protože je pořád co převádět
        "") ;;Ukončí rekurzy v dalším kole
      (str ret (tabulka mezi-klic))))
     ret)))

;;Komplet pravidla
;;0 = PN if předtím 00
;;0 = NN if předtím 10
;;1 = NP
(defn mfm-encode
  "Zakóduje binární text pomocí MFM modulace"
  ([bin-text] (mfm-encode (first bin-text) \0 (rest bin-text) nil))
  ([bit prev tail ret]
   (if (not= bit nil) ;;Dokud máme bity ke kontrole
     (recur
      (first tail) ;;Předáváme další bit na průzkum
      bit ;;Předchozí bude současný bit
      (rest tail) ;;Ocas bude ocas ocasu
      (str ret
           ;;Hlavní logika kde se rozhoduje o tom jaká modulace se použije
           (if (= bit \1)
             "NP"
             (if (= prev \0)
               "PN"
               "NN"))))
     ret)))

(defn fm-encode
  "Zakóduje text pomocí FN modulace"
  [bin-text]
  (cstr/replace
   (cstr/replace bin-text #"0" "PN")
   #"1" "PP"))

(defn -main
  "Vstupní bod, bere 1 argument - text k modulaci"
  [& args]
  (when (= nil args)
    (println "Použití: modulator [text k převedení]")
    (System/exit 1))
  (let [text-bin (nu/text-to-bin (first args))
        text-fm (fm-encode text-bin)
        text-mfm (mfm-encode text-bin)
        text-rll1 (rll-encode RLL-1 text-bin)
        text-rll2 (rll-encode RLL-2 text-bin)]
    (println "Text: " (first args))
    (println "   Bin: " text-bin)
    (println "    FM: " text-fm)
    (println "   MFM: " text-mfm)
    (println " RLL-1: " text-rll1)
    (println " RLL-2: " text-rll2)))
